File: 0496-next-greater-element-i/0496-next-greater-element-i.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var nextGreaterElement = function(nums1, nums2) {
    let stack=[];
    let map=new Map();

    for(num of nums2){
        while(stack.length>0 && num > stack[stack.length-1]){
            let smallest=stack.pop();
            map.set(smallest,num);
        }
        stack.push(num);
    }

    while(stack.length>0){
        map.set(stack.pop(),-1);
    }

    return nums1.map((val)=>map.get(val));
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stack = []
        nextGreater = {}

        for num in nums2:
            while stack and stack[-1] < num:   # while, not if
                smaller = stack.pop()
                nextGreater[smaller] = num
            stack.append(num)
        
        while stack:
            nextGreater[stack.pop()] = -1
        
        result=[]
        for num in nums1:
            result.append(nextGreater[num])
        return result    